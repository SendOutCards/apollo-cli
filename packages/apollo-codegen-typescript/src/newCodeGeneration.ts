import { BasicGeneratedFile } from 'apollo-codegen-core/lib/utilities/CodeGenerator';

import { CompilerContext } from 'apollo-codegen-core/lib/compiler';

import Printer from './printer';
import * as path from 'path';

import {
  importDeclarationForFragmentRawString,
  typeAliasDeclarationForFragment,
  typeAliasDeclarationForOperation,
  variableDeclarationForOperation,
  typeAliasDeclarationForGraphQLInputObjectType,
  variableDeclarationForGraphQLInputObjectType,
  enumDeclarationForGraphQLEnumType,
  exportDeclaration,
  importDeclarationForOperationRawString,
  globalImportDeclarationForFragmentDependencies
} from './types';
import { isInputObjectType, isEnumType } from 'graphql';

class TypescriptGeneratedFile implements BasicGeneratedFile {
  fileContents: string;

  constructor(fileContents: string) {
    this.fileContents = fileContents;
  }
  get output() {
    return this.fileContents;
  }
}

interface IGeneratedFileOptions {
  outputPath?: string;
  globalSourcePath?: string;
}

interface IGeneratedFile {
  sourcePath: string;
  fileName: string;
  content: (options?: IGeneratedFileOptions) => TypescriptGeneratedFile;
}

const globalImportPath = (outputPath: string, globalSourcePath: string) =>
  path.relative(
    path.dirname(outputPath),
    path.join(path.dirname(globalSourcePath), path.basename(globalSourcePath, '.ts'))
  );

export function generateLocalSource(context: CompilerContext): IGeneratedFile[] {
  return Object.values(context.fragments)
    .map(fragment => ({
      sourcePath: fragment.filePath,
      fileName: `${fragment.fragmentName}.ts`,
      content: (options?: IGeneratedFileOptions) => {
        return new TypescriptGeneratedFile(
          (options &&
            options.outputPath &&
            options.globalSourcePath &&
            `${fragment.filePath}\n${options.outputPath}\n${options.globalSourcePath}\n${globalImportPath(
              options.outputPath,
              options.globalSourcePath
            )}`) ||
            ''
        );
      }
    }))
    .concat(
      Object.values(context.operations).map(operation => ({
        sourcePath: operation.filePath,
        fileName: `${operation.operationName}.ts`,
        content: (options?: IGeneratedFileOptions) => {
          return new TypescriptGeneratedFile(
            (options &&
              options.outputPath &&
              options.globalSourcePath &&
              globalImportPath(options.outputPath, options.globalSourcePath)) ||
              ''
          );
        }
      }))
    );
}

export function generateGlobalSource(context: CompilerContext): TypescriptGeneratedFile {
  const printer = new Printer();
  context.typesUsed.forEach(type => {
    if (isEnumType(type)) {
      printer.enqueue(exportDeclaration(enumDeclarationForGraphQLEnumType(type)));
    } else if (isInputObjectType(type)) {
      printer.enqueue(exportDeclaration(typeAliasDeclarationForGraphQLInputObjectType(type)));
      printer.enqueue(exportDeclaration(variableDeclarationForGraphQLInputObjectType(type)));
    }
  });
  for (const key in context.fragments) {
    const fragment = context.fragments[key];
    printer.enqueue(importDeclarationForFragmentRawString(fragment));
    const globalImport = globalImportDeclarationForFragmentDependencies(fragment);
    if (globalImport) {
      printer.enqueue(globalImport);
    }
    printer.enqueue(exportDeclaration(typeAliasDeclarationForFragment(fragment)));
  }
  for (const key in context.operations) {
    const operation = context.operations[key];
    printer.enqueue(importDeclarationForOperationRawString(operation));
    printer.enqueue(exportDeclaration(typeAliasDeclarationForOperation(operation)));
    printer.enqueue(exportDeclaration(variableDeclarationForOperation(operation)));
  }
  const result = printer.print();
  return new TypescriptGeneratedFile(result);
}
