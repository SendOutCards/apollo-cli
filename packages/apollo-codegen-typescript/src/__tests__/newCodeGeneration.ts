import { parse } from 'graphql';

import { loadSchema } from 'apollo-codegen-core/lib/loading';
const schema = loadSchema(require.resolve('../../../common-test/fixtures/starwars/schema.json'));
const miscSchema = loadSchema(require.resolve('../../../common-test/fixtures/misc/schema.json'));

import { compileToIR, CompilerOptions, CompilerContext } from 'apollo-codegen-core/lib/compiler';

import { generateLocalSource, generateGlobalSource } from '../newCodeGeneration';

function compile(
  source: string,
  options: CompilerOptions = {
    mergeInFieldsFromFragmentSpreads: true,
    addTypename: false
  }
): CompilerContext {
  const document = parse(source);
  return compileToIR(schema, document, options);
}

function compileMisc(
  source: string,
  options: CompilerOptions = {
    mergeInFieldsFromFragmentSpreads: true,
    addTypename: false
  }
): CompilerContext {
  const document = parse(source);
  return compileToIR(miscSchema, document, options);
}

describe('Typescript newCodeGeneration', () => {
  test('simple hero query', () => {
    const context = compile(`
      query HeroName($episode: Episode) {
        hero(episode: $episode) {
          name
          id
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('simple mutation', () => {
    const context = compile(`
      mutation ReviewMovie($episode: Episode, $review: ReviewInput) {
        createReview(episode: $episode, review: $review) {
          stars
          commentary
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('simple fragment', () => {
    const context = compile(`
      fragment SimpleFragment on Character {
        name
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('fragment with fragment spreads', () => {
    const context = compile(`
      fragment SimpleFragment on Character {
        name
      }

      fragment AnotherFragment on Character {
        id
        ...SimpleFragment
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('fragment with fragment spreads with inline fragment', () => {
    const context = compile(`
      fragment SimpleFragment on Character {
        name
      }

      fragment AnotherFragment on Character {
        id
        ...SimpleFragment

        ... on Human {
          appearsIn
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('query with fragment spreads', () => {
    const context = compile(`
      fragment SimpleFragment on Character {
        name
      }

      query HeroFragment($episode: Episode) {
        hero(episode: $episode) {
          ...SimpleFragment
          id
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('inline fragment', () => {
    const context = compile(`
      query HeroInlineFragment($episode: Episode) {
        hero(episode: $episode) {
          ... on Character {
            name
          }
          id
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('inline fragment on type conditions', () => {
    const context = compile(`
      query HeroName($episode: Episode) {
        hero(episode: $episode) {
          name
          id

          ... on Human {
            homePlanet
            friends {
              name
            }
          }

          ... on Droid {
            appearsIn
          }
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('inline fragment on type conditions with differing inner fields', () => {
    const context = compile(`
      query HeroName($episode: Episode) {
        hero(episode: $episode) {
          name
          id

          ... on Human {
            homePlanet
            friends {
              name
            }
          }

          ... on Droid {
            appearsIn
            friends {
              id
            }
          }
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('fragment spreads with inline fragments', () => {
    const context = compile(`
      query HeroName($episode: Episode) {
        hero(episode: $episode) {
          name
          id
          ...HumanFragment
          ...DroidFragment
        }
      }

      fragment HumanFragment on Human {
        homePlanet
        friends {
          ... on Human {
            name
          }

          ... on Droid {
            id
          }
        }
      }

      fragment DroidFragment on Droid {
        appearsIn
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('handles multiline graphql comments', () => {
    const context = compileMisc(`
      query CustomScalar {
        commentTest {
          multiLine
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('multiple nested non-null list enum', () => {
    const context = compileMisc(`
      query nesting {
        nesting {
          propA
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('multiple nested list enum', () => {
    const context = compileMisc(`
      query nesting {
        nesting {
          propB
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });

  test('duplicates', () => {
    const context = compileMisc(`
      mutation duplicates($a: EnumCommentTestCase!, $b: EnumCommentTestCase!, $c: Duplicate!) {
        duplicates(a: $a, b: $b, c: $c) {
          propA
          propB
        }
      }
    `);

    const output = generateLocalSource(context).map(f => ({
      ...f,
      content: f.content({
        outputPath: '/some/file/ComponentA.tsx',
        globalSourcePath: '/__generated__/globalTypes.ts'
      })
    }));
    expect(output).toMatchSnapshot();
    expect(generateGlobalSource(context)).toMatchSnapshot();
  });
});
