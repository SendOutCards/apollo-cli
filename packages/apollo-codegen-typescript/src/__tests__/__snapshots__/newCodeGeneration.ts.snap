// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typescript newCodeGeneration duplicates 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "duplicates.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration duplicates 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum EnumCommentTestCase {
  first = \\"first\\",
  second = \\"second\\",
}

export type Duplicate = {
  propA: EnumCommentTestCase;
  propB: EnumCommentTestCase[];
};

export const Duplicate = (propA: EnumCommentTestCase, propB: EnumCommentTestCase[]) => ({
  propA,
  propB
});

import duplicatesRawString from \\"../GraphQL request\\";

export type duplicates = {
  duplicates: {
    propA: EnumCommentTestCase[][];
    propB: Maybe<Maybe<Maybe<EnumCommentTestCase>[]>[]>;
  };
};

export const duplicates = (a: EnumCommentTestCase, b: EnumCommentTestCase, c: Duplicate) => ({
  query: rawduplicates,
  variables: {
    a,
    b,
    c
  }
});
",
}
`;

exports[`Typescript newCodeGeneration fragment spreads with inline fragments 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "GraphQL request
/some/file/ComponentA.tsx
/__generated__/globalTypes.ts
../../__generated__/globalTypes",
    },
    "fileName": "HumanFragment.ts",
    "sourcePath": "GraphQL request",
  },
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "GraphQL request
/some/file/ComponentA.tsx
/__generated__/globalTypes.ts
../../__generated__/globalTypes",
    },
    "fileName": "DroidFragment.ts",
    "sourcePath": "GraphQL request",
  },
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "HeroName.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration fragment spreads with inline fragments 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum Episode {
  NEWHOPE = \\"NEWHOPE\\",
  EMPIRE = \\"EMPIRE\\",
  JEDI = \\"JEDI\\",
}

import humanFragmentRawString from \\"../GraphQL request\\";

import { Maybe } from \\"../__generated__/globalTypes\\";

export type HumanFragment = {
  homePlanet: Maybe<string>;
  friends: Maybe<Maybe<(If<\\"Human\\", {
    name: string;
  }> | If<\\"Droid\\", {
    id: string;
  }>)>[]>;
};

import droidFragmentRawString from \\"../GraphQL request\\";

import { Maybe, Episode } from \\"../__generated__/globalTypes\\";

export type DroidFragment = {
  appearsIn: Maybe<Episode>[];
};

import heroNameRawString from \\"../GraphQL request\\";

export type HeroName = {
  hero: Maybe<{
    name: string;
    id: string;
  } & (If<\\"Human\\", HumanFragment> | If<\\"Droid\\", DroidFragment>)>;
};

export const HeroName = (episode: Optional<Episode>) => ({
  query: rawHeroName,
  variables: {
    episode
  }
});
",
}
`;

exports[`Typescript newCodeGeneration fragment with fragment spreads 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "GraphQL request
/some/file/ComponentA.tsx
/__generated__/globalTypes.ts
../../__generated__/globalTypes",
    },
    "fileName": "SimpleFragment.ts",
    "sourcePath": "GraphQL request",
  },
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "GraphQL request
/some/file/ComponentA.tsx
/__generated__/globalTypes.ts
../../__generated__/globalTypes",
    },
    "fileName": "AnotherFragment.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration fragment with fragment spreads 2`] = `
TypescriptGeneratedFile {
  "fileContents": "import simpleFragmentRawString from \\"../GraphQL request\\";

export type SimpleFragment = {
  name: string;
};

import anotherFragmentRawString from \\"../GraphQL request\\";

export type AnotherFragment = SimpleFragment & {
  id: string;
};
",
}
`;

exports[`Typescript newCodeGeneration fragment with fragment spreads with inline fragment 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "GraphQL request
/some/file/ComponentA.tsx
/__generated__/globalTypes.ts
../../__generated__/globalTypes",
    },
    "fileName": "SimpleFragment.ts",
    "sourcePath": "GraphQL request",
  },
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "GraphQL request
/some/file/ComponentA.tsx
/__generated__/globalTypes.ts
../../__generated__/globalTypes",
    },
    "fileName": "AnotherFragment.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration fragment with fragment spreads with inline fragment 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum Episode {
  NEWHOPE = \\"NEWHOPE\\",
  EMPIRE = \\"EMPIRE\\",
  JEDI = \\"JEDI\\",
}

import simpleFragmentRawString from \\"../GraphQL request\\";

export type SimpleFragment = {
  name: string;
};

import anotherFragmentRawString from \\"../GraphQL request\\";

import { Maybe, Episode } from \\"../__generated__/globalTypes\\";

export type AnotherFragment = SimpleFragment & {
  id: string;
} & (If<\\"Human\\", {
  appearsIn: Maybe<Episode>[];
}> | If<\\"Droid\\", {}>);
",
}
`;

exports[`Typescript newCodeGeneration handles multiline graphql comments 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "CustomScalar.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration handles multiline graphql comments 2`] = `
TypescriptGeneratedFile {
  "fileContents": "import customScalarRawString from \\"../GraphQL request\\";

export type CustomScalar = {
  commentTest: Maybe<{
    multiLine: Maybe<string>;
  }>;
};

export const CustomScalar = () => ({
  query: rawCustomScalar,
  variables: {}
});
",
}
`;

exports[`Typescript newCodeGeneration inline fragment 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "HeroInlineFragment.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration inline fragment 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum Episode {
  NEWHOPE = \\"NEWHOPE\\",
  EMPIRE = \\"EMPIRE\\",
  JEDI = \\"JEDI\\",
}

import heroInlineFragmentRawString from \\"../GraphQL request\\";

export type HeroInlineFragment = {
  hero: Maybe<{
    name: string;
    id: string;
  }>;
};

export const HeroInlineFragment = (episode: Optional<Episode>) => ({
  query: rawHeroInlineFragment,
  variables: {
    episode
  }
});
",
}
`;

exports[`Typescript newCodeGeneration inline fragment on type conditions 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "HeroName.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration inline fragment on type conditions 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum Episode {
  NEWHOPE = \\"NEWHOPE\\",
  EMPIRE = \\"EMPIRE\\",
  JEDI = \\"JEDI\\",
}

import heroNameRawString from \\"../GraphQL request\\";

export type HeroName = {
  hero: Maybe<{
    name: string;
    id: string;
  } & (If<\\"Human\\", {
    homePlanet: Maybe<string>;
    friends: Maybe<Maybe<{
      name: string;
    }>[]>;
  }> | If<\\"Droid\\", {
    appearsIn: Maybe<Episode>[];
  }>)>;
};

export const HeroName = (episode: Optional<Episode>) => ({
  query: rawHeroName,
  variables: {
    episode
  }
});
",
}
`;

exports[`Typescript newCodeGeneration inline fragment on type conditions with differing inner fields 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "HeroName.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration inline fragment on type conditions with differing inner fields 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum Episode {
  NEWHOPE = \\"NEWHOPE\\",
  EMPIRE = \\"EMPIRE\\",
  JEDI = \\"JEDI\\",
}

import heroNameRawString from \\"../GraphQL request\\";

export type HeroName = {
  hero: Maybe<{
    name: string;
    id: string;
  } & (If<\\"Human\\", {
    homePlanet: Maybe<string>;
    friends: Maybe<Maybe<{
      name: string;
    }>[]>;
  }> | If<\\"Droid\\", {
    appearsIn: Maybe<Episode>[];
    friends: Maybe<Maybe<{
      id: string;
    }>[]>;
  }>)>;
};

export const HeroName = (episode: Optional<Episode>) => ({
  query: rawHeroName,
  variables: {
    episode
  }
});
",
}
`;

exports[`Typescript newCodeGeneration multiple nested list enum 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "nesting.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration multiple nested list enum 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum EnumCommentTestCase {
  first = \\"first\\",
  second = \\"second\\",
}

import nestingRawString from \\"../GraphQL request\\";

export type nesting = {
  nesting: {
    propB: Maybe<Maybe<Maybe<EnumCommentTestCase>[]>[]>;
  };
};

export const nesting = () => ({
  query: rawnesting,
  variables: {}
});
",
}
`;

exports[`Typescript newCodeGeneration multiple nested non-null list enum 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "nesting.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration multiple nested non-null list enum 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum EnumCommentTestCase {
  first = \\"first\\",
  second = \\"second\\",
}

import nestingRawString from \\"../GraphQL request\\";

export type nesting = {
  nesting: {
    propA: EnumCommentTestCase[][];
  };
};

export const nesting = () => ({
  query: rawnesting,
  variables: {}
});
",
}
`;

exports[`Typescript newCodeGeneration query with fragment spreads 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "GraphQL request
/some/file/ComponentA.tsx
/__generated__/globalTypes.ts
../../__generated__/globalTypes",
    },
    "fileName": "SimpleFragment.ts",
    "sourcePath": "GraphQL request",
  },
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "HeroFragment.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration query with fragment spreads 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum Episode {
  NEWHOPE = \\"NEWHOPE\\",
  EMPIRE = \\"EMPIRE\\",
  JEDI = \\"JEDI\\",
}

import simpleFragmentRawString from \\"../GraphQL request\\";

export type SimpleFragment = {
  name: string;
};

import heroFragmentRawString from \\"../GraphQL request\\";

export type HeroFragment = {
  hero: Maybe<SimpleFragment & {
    id: string;
  }>;
};

export const HeroFragment = (episode: Optional<Episode>) => ({
  query: rawHeroFragment,
  variables: {
    episode
  }
});
",
}
`;

exports[`Typescript newCodeGeneration simple fragment 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "GraphQL request
/some/file/ComponentA.tsx
/__generated__/globalTypes.ts
../../__generated__/globalTypes",
    },
    "fileName": "SimpleFragment.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration simple fragment 2`] = `
TypescriptGeneratedFile {
  "fileContents": "import simpleFragmentRawString from \\"../GraphQL request\\";

export type SimpleFragment = {
  name: string;
};
",
}
`;

exports[`Typescript newCodeGeneration simple hero query 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "HeroName.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration simple hero query 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum Episode {
  NEWHOPE = \\"NEWHOPE\\",
  EMPIRE = \\"EMPIRE\\",
  JEDI = \\"JEDI\\",
}

import heroNameRawString from \\"../GraphQL request\\";

export type HeroName = {
  hero: Maybe<{
    name: string;
    id: string;
  }>;
};

export const HeroName = (episode: Optional<Episode>) => ({
  query: rawHeroName,
  variables: {
    episode
  }
});
",
}
`;

exports[`Typescript newCodeGeneration simple mutation 1`] = `
Array [
  Object {
    "content": TypescriptGeneratedFile {
      "fileContents": "../../__generated__/globalTypes",
    },
    "fileName": "ReviewMovie.ts",
    "sourcePath": "GraphQL request",
  },
]
`;

exports[`Typescript newCodeGeneration simple mutation 2`] = `
TypescriptGeneratedFile {
  "fileContents": "export enum Episode {
  NEWHOPE = \\"NEWHOPE\\",
  EMPIRE = \\"EMPIRE\\",
  JEDI = \\"JEDI\\",
}

export type ReviewInput = {
  stars: number;
  commentary: Optional<string>;
  favorite_color: Optional<ColorInput>;
};

export const ReviewInput = (stars: number, commentary: Optional<string>, favorite_color: Optional<ColorInput>) => ({
  stars,
  commentary,
  favorite_color
});

export type ColorInput = {
  red: number;
  green: number;
  blue: number;
};

export const ColorInput = (red: number, green: number, blue: number) => ({
  red,
  green,
  blue
});

import reviewMovieRawString from \\"../GraphQL request\\";

export type ReviewMovie = {
  createReview: Maybe<{
    stars: number;
    commentary: Maybe<string>;
  }>;
};

export const ReviewMovie = (episode: Optional<Episode>, review: Optional<ReviewInput>) => ({
  query: rawReviewMovie,
  variables: {
    episode,
    review
  }
});
",
}
`;
